File: Programas/NoInformados/Rompecabezas/1. Busqueda en Profundidad v1.0.0/JAVA DFS/DFS.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;

public class DFS {

    // Recupera los nodos visitados para llegar al nodo final
    private static ArrayList<Rompecabezas> trayectoria( Rompecabezas nodoFinal ){
        ArrayList<Rompecabezas> secuencia = new ArrayList<>();
        secuencia.add( nodoFinal );
        while( nodoFinal.nodoPadre != null ){
            nodoFinal = nodoFinal.nodoPadre;
            secuencia.add( nodoFinal );
        }
        Collections.reverse(secuencia);
        return secuencia;
    }

    // Algoritmo Busqueda en Profundidad
    public static ArrayList<Rompecabezas> Bucar( Rompecabezas nodoInicial , Rompecabezas nodoFinal ){
        ArrayList<Rompecabezas> agenda = new ArrayList<>();
        HashSet<Long> configuracionesExploradas = new HashSet<>();

        if( nodoFinal.compararCeldas(nodoInicial) )
            return trayectoria(nodoInicial);
        
        agenda.add( nodoInicial );

        while( agenda.size() > 0 ){
            Rompecabezas nodo = agenda.remove( agenda.size()-1 );

            configuracionesExploradas.add( nodo.tablero );
            for( Rompecabezas hijo : nodo.expandirNodos() ){
                if( nodoFinal.compararCeldas(hijo) )
                    return trayectoria( hijo );
                else if( !configuracionesExploradas.contains( hijo.tablero ) )
                    agenda.add( hijo );
            }
        }

        return null;
    }
}

File: Programas/NoInformados/Rompecabezas/1. Busqueda en Profundidad v1.0.0/JavaScript DFS/DFS.js
// Recupera los nodos visitados para llegar al nodo final
function trayectoria( nodoFinal ){
    let secuencia = []
    secuencia.push( nodoFinal )
    while( nodoFinal.nodoPadre ){
        nodoFinal = nodoFinal.nodoPadre
        secuencia.push( nodoFinal )
    }
    secuencia.reverse()
    return secuencia
}

// Algoritmo Busqueda en Profundidad
export class DFS {
    static Buscar(nodoInicial, nodoFinal) {

        let agenda = []
        let configuracionesExploradas = new Set()

        if (nodoFinal.compararCeldas(nodoInicial))
            return trayectoria( nodoInicial )

        agenda.push(nodoInicial)

        while (agenda.length > 0) {

            let nodo = agenda.pop()

            configuracionesExploradas.add(nodo.celdas)

            for (let hijo of nodo.expandir()) {
                if (nodoFinal.compararCeldas(hijo))
                    return trayectoria(hijo)
                else if (!configuracionesExploradas.has(hijo.celdas))
                    agenda.push(hijo)
            }
        }

        return null
    }
}